hibernate-core/src/main/java/org/hibernate/query/hql/internal/QuerySplitter.java

```diff
@@ -1037,6 +1381,27 @@ public SqmCreationOptions getCreationOptions() {
 			return () -> false;
 		}
 
+		public <X extends SqmFrom<?, ?>> X findSqmFromCopy(SqmFrom sqmFrom) {
+			SqmFrom copy = sqmFromCopyMap.get( sqmFrom );
+			if ( copy != null ) {
+				return (X) copy;
+			}
+			else {
+				return (X) sqmFrom.accept( this );
+			}
+		}
+
+		private void visitJoins(SqmFrom<?, ?> join, SqmAttributeJoin copy) {
```

Comment:
Nice refactoring. I think you should be able to use that for all joins though i.e. `visitCrossJoin` etc.

Also, what about fetches and treats? Do we not support that in polymorphic queries? Might be worth adding an assert statement here then at least to make this clear.

Reply:
done, fI added a test for fetches and they seems working, for treat I added an exception